### ISSUE TITLE
E0308 diagnostic false positive involving macro use in closure

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-analyzer#23038]
Using the [`locale!()`](https://docs.rs/icu/latest/icu/locale/macro.locale.html) macro from the `icu` crate seems to cause issues with type inference involving coercion: the code snippet below produces "expected fn(), found {closure#...}" but is accepted by `rustc`. The macro uses crate internals and I was not sure how to simplify the example further.

**rust-analyzer version**: rust-analyzer version: 0.3.2997-standalone

**rustc version**: rustc 1.97.1 (8bab26f4f 2026-07-14)

**editor or extension**: VSCode on WSL (Ubuntu)

**relevant settings**: experimental diagnostics are disabled.

**code snippet to reproduce**:
```rust
use icu::locale::locale;

fn foo(a: fn()) {}

fn main() {
    foo(|| {
        locale!("en-US");
    });
}
```

**Comment by ChayimFriedman2** [comment]
Minimized example:
```rust
fn foo(_: fn()) {}

fn main() {
    foo(|| {
        const {
            match Some(()) {
                Some(v) => v,
                None => (),
            }
        }
    });
}
```
A little bit more:
```rust
fn foo(_: fn()) {}

fn main() {
    foo(|| {
        const {
            let v = ();
            v
        }
    });
}
```
